src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -337,75 +337,112 @@ public static unsafe void ClearWithReferences(ref IntPtr ip, nuint pointerSizeLe
-            {
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -1) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -2) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -3) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -4) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -5) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -6) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -7) = default;
-                Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -8) = default;
-            }
-
-            Debug.Assert(pointerSizeLength <= 7);
-
-            // The logic below works by trying to minimize the number of branches taken for any
-            // given range of lengths. For example, the lengths [ 4 .. 7 ] are handled by a single
-            // branch, [ 2 .. 3 ] are handled by a single branch, and [ 1 ] is handled by a single
-            // branch.
-            //
-            // We can write both forward and backward as a perf improvement. For example,
-            // the lengths [ 4 .. 7 ] can be handled by zeroing out the first four natural
-            // words and the last 3 natural words. In the best case (length = 7), there are
-            // no overlapping writes. In the worst case (length = 4), there are three
-            // overlapping writes near the middle of the buffer. In perf testing, the
-            // penalty for performing duplicate writes is less expensive than the penalty
-            // for complex branching.
-
-            if (pointerSizeLength >= 4)
-            {
-                goto Write4To7;
-            }
-            else if (pointerSizeLength >= 2)
-            {
-                goto Write2To3;
-            }
-            else if (pointerSizeLength > 0)
-            {
-                goto Write1;
-            }
-            else
+            do
             {
-                return; // nothing to write
-            }
-
-        Write4To7:
-            Debug.Assert(pointerSizeLength >= 4);
-
-            // Write first four and last three.
-            Unsafe.Add(ref ip, 2) = default;
-            Unsafe.Add(ref ip, 3) = default;
-            Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -3) = default;
-            Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -2) = default;
-
-        Write2To3:
-            Debug.Assert(pointerSizeLength >= 2);
-
-            // Write first two and last one.
-            Unsafe.Add(ref ip, 1) = default;
-            Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -1) = default;
-
-        Write1:
-            Debug.Assert(pointerSizeLength >= 1);
+                // This switch is expected to be lowered to a jump table
```

Comment:
This works great for microbenchmarks, but it works poorly for real work loads where the jump target is poorly predicated. We used to have switch table like this in managed memcopy originally, but it was later removed.

Reply:
> This works great for microbenchmarks, but it works poorly for real work loads where the jump target is poorly predicated. We used to have switch table like this in managed memcopy originally, but it was later removed.

Should I revert to the previous shape? I was mostly interested in aligning to 16 bytes (it's the reason I am seeing nice results on arm64) and slightly different addressing mode in the main loop, to make it foldable into `stp`

Reply:
Btw, presumably, PGO can re-shuffle the switch (or convert to if-else) based on profile

Reply:
> Should I revert to the previous shape?

Yes, unless you can prove that the switch with jump table is better for real workloads.

For reference, https://github.com/dotnet/coreclr/pull/9786 is the PR that got rid of the switch from memory copy. It went through extensive validation.

Reply:
> This works great for microbenchmarks, but it works poorly for real work loads where the jump target is poorly predicated. We used to have switch table like this in managed memcopy originally, but it was later removed.

@jkotas, I'm not sure what you mean here, could you elaborate?

The hardware optimization manuals and underlying native memcpy algorithms across all 3 major implementations (MSVC, GCC, Clang; for x64 and Arm64) use a jump table explicitly for this reason. Because an unconditional and unpredictable branch, is better than a tight tree of conditional and still unpredictable branches.

The premise is that a fallthrough jump table allows a tight, unrolled version of handling for the trailing elements to exist. This then requires 1 unpredicted branch to decide which of the immediately following and equidistant switch cases needs to be jumped to. These cases are often already in the instruction/decode cache due to being in the same or following 32-byte code window.

Reply:
https://github.com/dotnet/coreclr/pull/9786 was done by Intel engineers to address bottlenecks observed in real workloads. Yes, it goes contrary to the guidance you have quoted.

Possible explanation: Conditional jumps and indirect jumps use two different predictors. The relative performance of conditional jumps vs. indirect jump depends on relative pressure on each predictor. .NET code tends to have more indirect jumps than an average code. It means that the indirect jump predictor sees more pressure in .NET code and it is more profitable to depend on indirect jump predictor in .NET memcpy implementation.